src/lib.rs

```diff
@@ -813,26 +967,54 @@ where
       return Err(NovaError::ProofVerifyError);
     }
 
-    // fold the secondary's running instance with the last instance to get a folded instance
-    let f_U_secondary = self.nifs_secondary.verify(
+    // fold secondary U/W with secondary u/w to get Uf/Wf
+    let r_Uf_secondary = self.nifs_Uf_secondary.verify(
       &vk.ro_consts_secondary,
       &scalar_as_base::<E1>(vk.pp_digest),
       &self.r_U_secondary,
       &self.l_u_secondary,
     )?;
 
+    // fold Uf/Wf with random inst/wit to get U1/W1
+    let r_Un_secondary = self.nifs_Un_secondary.verify_relaxed(
```

Comment:
There seems to be a bit of inconsistency here. The code that corresponds to this in the prover is factored out into a separate method, but here, it is inlined right?

Reply:
Yes, you're right, inconsistent - an artifact of having separate functionality for RecursiveSNARK to prove and verify it's own random layer (before we removed that). I should either separate the first part of verification into a separate function, or combine the proving.

The downside of separating out verification is that it is not a "full verification" (i.e. we will not call `is_sat` at the end, as we would have if we were just verifying IVC proof). (Proving is "full".) I guess this doesn't really matter, but is a different sort of inconsistency.

But separating out functionality is nice and modular ... what do you prefer?

Reply:
maybe if both functions are in `RecursiveSNARKRandomized`, which is private to the user (only called inside of CompressedSNARK methods), it doesn't matter if it's not full verification.

Reply:
> Yes, you're right, inconsistent - an artifact of having separate functionality for RecursiveSNARK to prove and verify it's own random layer (before we removed that). I should either separate the first part of verification into a separate function, or combine the proving.
> 
> The downside of separating out verification is that it is not a "full verification" (i.e. we will not call `is_sat` at the end, as we would have if we were just verifying IVC proof). (Proving is "full".) I guess this doesn't really matter, but is a different sort of inconsistency.
> 
> But separating out functionality is nice and modular ... what do you prefer?

Since we are not anyway allowing RecursiveSNARK objects that are randomized to be externally used. Perhaps, for this PR, we could just inline the randomization process inside CompressedSNARK's prove method (analogous to how the verify method verifies the randomization). This would also avoid maintaining an additional type right?

Reply:
yes, that sounds good.